### PR TITLE
Breaking: Switch to ESM (fixes #70)

### LINKED ISSUE #70: Convert package to ESM
To convert this package from CommonJS to ECMAScript modules (ESM):

- [ ] Add `type: module` to `package.json`
- [ ] Convert files in `lib` to ESM
- [ ] Convert files in `tests/lib` to ESM
- [ ] Upgrade Mocha
- [ ] Create `build` script in `package.json` to generate CommonJS file (`.cjs`) in `dist` directory (use Rollup)
- [ ] Add `dist` to `.gitignore`
- [ ] Run `build` in `prepare` script in `package.json` to ensure build is run at key moments
- [ ] Update `package.json` so that `main` points in the CJS file in `dist` and `exports` provides both `require` and `import` options
- [ ] Update README instructions

### PR DESCRIPTION
**Pull request by brettz9**
Breaking: Switch to ESM

Fixes #70.

### CODE REVIEW

**Comment by brettz9** [comment]
Am I still getting commit message errors because I've more than one commit in the PR?

**Comment by nzakas** [comment]
Yes, when there's more than one commit then it checks the pull request title. I fixed it.

**Review by nzakas — CHANGES_REQUESTED** [review]
We will need to update the Node.js testing matrix to match the 12/14/16 that we have in Espree and `eslint-visitor-keys` to get everything working.

Thanks again for taking this on!

File: lib/definition.js
Comment (brettz9) [reply lib/definition.js:28]: Yeah, it's the current enforcement by `eslint-plugin-jsdoc`. The preference can be overridden in your eslint config.

**Comment by brettz9** [comment]
> We will need to update the Node.js testing matrix to match the 12/14/16 that we have in Espree and `eslint-visitor-keys` to get everything working.

Updated.

> Thanks again for taking this on!

My pleasure. Really like ESM both for it being a universal syntax and conducing to polyglot browser-friendly code. Also eager to get closer to using ESM for plugins, etc.

File: Makefile.js
Comment (brettz9) [reply Makefile.js:24]: That was preexisting code, related to `shelljs/make.js`. As per the new explanatory comment just added:

```js
// `shelljs/make.js` global command to unset any `set('-e')` (to exit upon
//    first error)
```

File: lib/definition.js
Comment (brettz9) [reply lib/definition.js:28]: Sorry, I was mistaken about it being the new default. It appears your config must be changing it. And frankly, I agree with the change. (We currently prefer the tag name not declared as an alias by jsdoc.)

FWIW, what follows are my reasons for favoring your current config's approach of favoring `@constructor`...

Those `@class`/`@constructor` tags should only be necessary when jsdoc can't detect the surrounding type and name. In all the cases in `eslint-scope`, it appears straightforward, so I think the whole tag is redundant. (Even [`@classdesc`](https://jsdoc.app/tags-classdesc.html), which allows an accompanying description, would be redundant here since one could just describe the class as free form text of the block comment when the fact of it being a class and its name were both auto-discoverable from context.)

So, the `function aConstructorFunction () { this.a = 5; }` type of "class" would be the only one needing the tag, but I think that `@class` seemed to muddy the waters with ES6 classes whereas `@constructor` fits.

`@class` therefore might only be useful iI think n a situation where one wanted to document a variable as being a class outside of the context where it was defined, and for those wishing to do so, they'd have to be set in `settings.jsdoc.tagNamePreference`. See the [jsdoc/check-tag-names](https://github.com/gajus/eslint-plugin-jsdoc/#eslint-plugin-jsdoc-rules-check-tag-names).

File: Makefile.js
Comment (nzakas) [reply Makefile.js:24]: Thanks

File: lib/definition.js
Comment (nzakas) [reply lib/definition.js:28]: Ah ok, let's stick with what the config says. No need to fight another decision.

**Comment by nzakas** [comment]
I think this looks good. I'd like to merge #69 first and then rebase this on top of that before merging.

**Comment by nzakas** [comment]
#69 has been merged, so when you feel up to rebase on top of that, we can look at this again.

**Comment by brettz9** [comment]
Rebasing is done, including README as well as API and ESM adjustments.

File: lib/index.js
Comment (mdjermanovic) [inline lib/index.js:50]: This will crash our [eslint demo](https://eslint.org/demo) build.

I tried to copy this eslint-scope version over the one in `node-modules` under [src/js/demo](https://github.com/eslint/website/tree/master/src/js/demo), and then `npm run build:webpack` failed with:

```
ERROR in ./src/js/demo/node_modules/eslint-scope/dist/eslint-scope.cjs 5:9-22
Module not found: Error: Can't resolve 'fs' in '...\src\js\demo\node_modules\eslint-scope\dist'
```

File: lib/pattern-visitor.js
Comment (mdjermanovic) [inline lib/pattern-visitor.js:27]: `estraverse` provides only cjs export. I think we should use only import default from `estraverse` because [interoperability with CommonJS](https://nodejs.org/api/esm.html#esm_interoperability_with_commonjs) doesn't guarantee named exports:

> When importing CommonJS modules, the module.exports object is provided as the default export. Named exports may be available, provided by static analysis as a convenience for better ecosystem compatibility.

The same for `esrecurse` and `@typescript-eslint/parser`.

File: README.md
Comment (nzakas) [inline README.md:45]:
```suggestion
    leave(node, parent) {
```

Removing extra space before parens

File: lib/index.js
Comment (nzakas) [inline lib/index.js:58]: Just double checking - it looks like Scope was imported just to be exported again. Is it actually safe to remove it completely?

File: lib/index.js
Comment (nzakas) [inline lib/index.js:58]: As @mdjermanovic mentioned, this change won’t work in a browser, where ESLint currently works for our demo (and probably for others too). If version is needed, we may need to do something like what we did in the Espree repo to work around this limitation.

**Review by nzakas — CHANGES_REQUESTED** [review]

File: lib/pattern-visitor.js
Comment (brettz9) [reply lib/pattern-visitor.js]: Ok, I've switched to the default import syntax for these CJS modules.

File: lib/index.js
Comment (brettz9) [reply lib/index.js:58]: It appears I had missed that only `Scope` was being re-exported, so I can return to that basic approach if you wish (i.e., to stop re-exporting the other `Scope` subclasses and only re-export `Scope`).

As to whether `Scope` itself can be dropped, I don't personally know enough on how the package is being used, so I guess that may be a question for others here. But I can mention that the current `exports` would not allow access to the `scope.js` submodule.

File: lib/index.js
Comment (brettz9) [reply lib/index.js]: I have made a change akin to that in Espree. Let me know if you want to drop the `version` export entirely. We are currently exporting `package.json` in `exports` so the user should be able to get at it.

File: lib/index.js
Comment (brettz9) [reply lib/index.js]: I have removed the `fs` dependency, but currently the Rollup config I added is treating the following as external and is not bundling them: `assert`, `estraverse`, and `esrecurse`.

Since you use `external` in [espree](https://github.com/eslint/espree/blob/main/rollup.config.js), I figured you were leaving it to consumers to bundle, but if the demo is still not working after dropping `fs` and you want, I could add `@rollup/plugin-commonjs`, etc. to have the latter two bundled (with the disadvantage that they will be baked-in versions) and add something like [`rollup-plugin-unassert`](https://github.com/unassert-js/rollup-plugin-unassert) or [`rollup-plugin-node-builtins`](https://github.com/calvinmetcalf/rollup-plugin-node-builtins) to handle `assert`. Otherwise, I'm not sure how to proceed.

Comment (mdjermanovic) [reply lib/index.js]: > I have removed the `fs` dependency, but currently the Rollup config I added is treating the following as external and is not bundling them: `assert`, `estraverse`, and `esrecurse`.

That should be fine, we don't intend to bundle any external modules.

The currently published eslint-scope is a set of cjs modules from the `lib/` directory, so consumers already handle `require("assert")`, `require("estraverse")`, and `require("estraverse")` while bundling (e.g., eslint demo uses `node-polyfill-webpack-plugin` to handle `assert`). I think the new elint-scope will be basically the same thing in that regard, just packed in one module `dist/eslint-scope.cjs`.

**Comment by mdjermanovic** [comment]
It seems that `@babel-eslint/parser` uses some internal modules that won't be exported now, so this change will break it.

https://github.com/babel/babel/blob/9d620c2d423cec4cf5d555d4b8884288068bd17d/eslint/babel-eslint-parser/src/analyze-scope.cjs#L1-L4

```js
const escope = require("eslint-scope");
const { Definition } = require("eslint-scope/lib/definition");
const OriginalPatternVisitor = require("eslint-scope/lib/pattern-visitor");
const OriginalReferencer = require("eslint-scope/lib/referencer");
```

**Comment by mdjermanovic** [comment]
Webpack imports [lib/referencer](https://github.com/webpack/webpack/blob/c181294865dca01b28e6e316636fef5f2aad4eb6/lib/optimize/ConcatenatedModule.js#L9) directly, though this probably won't be necessary for them with an eslint-scope version that supports class fields.

File: lib/index.js
Comment (mdjermanovic) [reply lib/index.js]: `version` is used in `vue-eslint-parser` [here](https://github.com/vuejs/vue-eslint-parser/blob/66d424ecc108c8850c4ff225fea0199cfc8eee6b/src/common/eslint-scope.ts#L18)

File: lib/index.js
Comment (mdjermanovic) [reply lib/index.js:58]: `Scope` is used in `@angular-eslint/template-parser` [here](https://github.com/angular-eslint/angular-eslint/blob/a4beae803e48197b9897eb9a9742186dc18aa4b5/packages/template-parser/src/index.ts#L5)

**Comment by mdjermanovic** [comment]
`@typescript-eslint/experimental-utils`, which seems to be used by a lot of [plugins](https://www.npmjs.com/browse/depended/@typescript-eslint/experimental-utils), imports some modules directly:

* [lib/pattern-visitor](https://github.com/typescript-eslint/typescript-eslint/blob/1c1b572c3000d72cfe665b7afbada0ec415e7855/packages/experimental-utils/src/ts-eslint-scope/PatternVisitor.ts#L1)
* [lib/scope-manager](https://github.com/typescript-eslint/typescript-eslint/blob/1c1b572c3000d72cfe665b7afbada0ec415e7855/packages/experimental-utils/src/ts-eslint-scope/ScopeManager.ts#L1)
* [lib/referencer](https://github.com/typescript-eslint/typescript-eslint/blob/1c1b572c3000d72cfe665b7afbada0ec415e7855/packages/experimental-utils/src/ts-eslint-scope/Referencer.ts#L2)
* [lib/scope](https://github.com/typescript-eslint/typescript-eslint/blob/b1d44497ddf8ef4615c727fe4241e54f7f1aba65/packages/experimental-utils/src/ts-eslint-scope/Scope.ts#L3-L15)

File: lib/index.js
Comment (mdjermanovic) [reply lib/index.js:58]: Subclasses are imported in `@typescript-eslint/experimental-utils` [here](https://github.com/typescript-eslint/typescript-eslint/blob/b1d44497ddf8ef4615c727fe4241e54f7f1aba65/packages/experimental-utils/src/ts-eslint-scope/Scope.ts#L3-L15) (directly from `lib/scope` ). I'm not sure if they're actually used or it's just for types.

Comment (nzakas) [reply lib/index.js:58]: It seems like we should leave Scope as an export here.

Comment (brettz9) [reply lib/index.js:58]: Re: leaving Scope, how about the subclasses? If you want those, I can either re-export them too as I am now or we'd have to expand `exports` to allow access to the Scope file.

**Comment by nzakas** [comment]
@bradzacher can you comment on how typescript-eslint is using subpath exports from eslint-scope? We were planning on removing them when we discovered usage in typescript-eslint. Are these necessary or just convenient?

**Comment by nzakas** [comment]
@snitin315 do you know who on Webpack we could talk to about how they are using eslint-scope?

**Comment by snitin315** [comment]
> @snitin315 do you know who on Webpack we could talk to about how they are using eslint-scope?

@nzakas you can talk with Tobias, I found only this instance of `eslint-scope` https://github.com/webpack/webpack/blob/main/lib/optimize/ConcatenatedModule.js#L8

/cc @sokra

**Comment by bradzacher** [comment]
We wrap the module and provide types for it that are compatible with the TS AST types.
It's a nice-to-have as it enables people to use `eslint-scope` with the types without any manual weird type casts.
Though now-a-days we don't *really* need it as we have our own [scope analyser which is a fork of this package](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/scope-manager) which people should really use instead (because it supports TS features.. which you're going to need if you're using the TS AST types).

So yeah - could be breaking-change removed.

File: lib/index.js
Comment (mdjermanovic) [reply lib/index.js:58]: > or we'd have to expand `exports` to allow access to the Scope file.

If we decide to export some modules as subpaths (I guess we won't, but just in case), then we should also build their cjs equivalents in `dist/`.

**Comment by nzakas** [comment]
@bradzacher oh nice, thanks for the details! We will proceed with the breaking changes then.

@sokra it looks like Webpack is only using `Referencer` to monkey-patch in `PropertyDefinition` here: https://github.com/webpack/webpack/blob/main/lib/optimize/ConcatenatedModule.js#L67. Given that `eslint-scope` will now support class properties natively, do you have any other need to access `Referencer` directly? We'd like to remove it from the public API.

**Comment by sokra** [comment]
> Given that `eslint-scope` will now support class properties natively, do you have any other need to access `Referencer` directly? We'd like to remove it from the public API.

It's only a workaround for a bug. If you fix the bug, we don't need that exposed anymore.

**Comment by sokra** [comment]
webpack won't be able to use the ESM version anyway yet, since we still support node.js 10 (and other node.js version that doesn't have ESM support) and will only be able to drop that in the next major version.

**Comment by nzakas** [comment]
@sokra this package will still export a CommonJS entrypoint, so hopefully that will help.

File: lib/index.js
Comment (nzakas) [reply lib/index.js:58]: I don't think we need to provide any subpaths at this point (given latest responses from typescript-eslint and Webpack).

@brettz9 given that these are internal-only, my vote is to just export the same things we've already been exporting so we aren't changing the overall structure too much. If we aren't already exporting subclasses, I don't think there's a good reason to do so.

**Comment by nzakas** [comment]
Where we are: it seems that the subpaths being used by typescript-eslint and Webpack aren't required going forward, so we don't need to provide any subpath exports for those.

**Comment by nzakas** [comment]
@nicolo-ribaudo it looks like Babel’s ESLint parser is reaching into eslint-scope to access some undocumented APIs that will no longer be accessible in the next release. Can you help us understand what the code is doing and how removing access to these APIs will affect you?

https://github.com/babel/babel/blob/9d620c2d423cec4cf5d555d4b8884288068bd17d/eslint/babel-eslint-parser/src/analyze-scope.cjs#L1-L4

File: lib/index.js
Comment (brettz9) [reply lib/index.js:58]: Ok, sure. I've added a commit so it is back to just exporting `Scope`.

**Comment by sokra** [comment]
> @sokra this package will still export a CommonJS entrypoint, so hopefully that will help.

That helps. As long `main` points to the commonjs version, old node.js versions are happy.

Also we our great guide for the exports field: https://webpack.js.org/guides/package-exports/

**Comment by bradzacher** [comment]
@nzakas on behalf of the babel folks as I'm pretty sure I know what they are doing.

They're extending this scope analyser to add support for the additional syntax they support.

We at ts-eslint used to do the same thing, until we bit the bullet and forked/rewrote it due to our requirement to support the separate value+type scope analysis.

Babel does technically support TS, though the vast, vast majority of users use ts-eslint for that usecase. So instead babel-eslint serves people who are using pure JS with additional features not yet supported by espree / this project. Meaning their extension is usually fairly limited in comparison.

If you remove that access then it becomes hard for them to extend this package to add support for additional syntax without maintaining a complete fork of the project.

They can correct me if I'm wrong though :)

**Comment by nicolo-ribaudo** [comment]
Thanks @bradzacher, it's 100% correct.

@nzakas If those internal APIs will be made inaccessible from the outside, we will probably fork `eslint-scope` and maintain those visitors ourselves. However, it will be more likely for us to accidentally fall behind ESLint causing potential problems for `@babel/eslint-parser` users.

(cc @hzoo who has more experience than me in our ESLint integration)

**Comment by hzoo** [comment]
Not much else to add to what @bradzacher/@nicolo-ribaudo said, this is some code we've had for a while, (this was the last PR https://github.com/babel/babel-eslint/pull/542) but the general idea has been there since the beginning to monkey patch /updating scopes for new ast features/types/differences in the ast.

**Comment by nzakas** [comment]
@nicolo-ribaudo @hzoo Hmmm, okay. I don't think this is a great approach (any more than using undocumented APIs is ever a good idea), but we definitely don't want to break you.

Long-term I think we need to reimagine how this package works. If everyone keeps using it as a base to extend, then we should really design an API that allows for that in an official way rather than coming across one-off uses that are using private methods.

**TSC Summary:** This PR switches eslint-scope to be an ESM package. We had originally planned on removing access to APIs under `lib/`, but that will break `@babel/eslint-parser`.

**TSC Question:** How shall we proceed?

So I think we have to decide how to move forward, which I see as one of:

1. "Bless" these APIs and just export them directly from this package.
2. Create a `use-at-your-own-risk` entrypoint like we did for ESLint that exports the things we don't want to officially support.
3. Keep all of the current APIs available at the current locations.

**Comment by btmills** [comment]
Good news: in the most recent TSC meeting, we agreed to export the APIs under `lib/` as public APIs. (Option 1 above.)

**Comment by brettz9** [comment]
We are currently giving both `import` and `require` options for the main `lib/index.js` export, but does this new decision mean we need to add a Rollup process for each of the subfiles, or is the idea just to expose those `import` style and let consumers do any bundling/transpiling to CJS themselves?

**Comment by brettz9** [comment]
I.e., adding 16 vs. 8 lines for each one of the additional `lib` files (or 14 vs. 7 if dropping `version.js` since it can be obtained by the `package.json` which is already ready to be exported).

Also, of those files, the exports might be reduced further if those already being exported in full by `index.js` are removed, i.e., `scope-manager.js`, `reference.js`, and `variable.js`.

**Comment by mdjermanovic** [comment]
> We are currently giving both `import` and `require` options for the main `lib/index.js` export, but does this new decision mean we need to add a Rollup process for each of the subfiles, or is the idea just to expose those `import` style and let consumers do any bundling/transpiling to CJS themselves?

Hmm, my understanding of option 1 was that `lib/index.js` will re-export `Referencer`, `PatternVisitor`, and `Definition`, but I could be wrong.

**Comment by nzakas** [comment]
> Hmm, my understanding of option 1 was that lib/index.js will re-export Referencer, PatternVisitor, and Definition, but I could be wrong.

This is correct. We would like lib/index.JS to export ScopeManager, Scope, Referencer, Definition, etc. (Everything Babel/eslint-parser is using) So we would still end up with one file created by Rollup and you shouldn’t need to update the Rollup script at all.

**Comment by brettz9** [comment]
Ah, ok, that makes more sense :-)  Added a commit with the exports...

**Review by nzakas — APPROVED** [review]

**Review by nzakas — APPROVED** [review]
LGTM. Thanks for all our your work on this.

File: package.json
Comment (mdjermanovic) [inline package.json:24]: A problem with this is that our release process doesn't commit `lib/version.js`, so after the `6.0.0` release we'll still have `const version = "5.1.1";` in `lib/version.js`.

Comment (mdjermanovic) [reply package.json:24]: This would then cause the next release (e.g. `6.1.0`) to fail: `npm install` will trigger the `prepare` script, which will overwrite `lib/version.js` with `6.0.0`. Then, [npm version](https://github.com/eslint/eslint-release/blob/0fdda4da083887abfb7b11a3b1ed8e82c4d3d99e/lib/release-ops.js#L383) will throw error "Git working directory not clean".

File: Makefile.js
Comment (mdjermanovic) [inline Makefile.js:98]:
```suggestion
    let lastReturn = exec(`${NODE} ${MOCHA} -- -R progress -c ${CJS_TEST_FILES}`);
```

To make it work on Windows.

File: package.json
Comment (nzakas) [reply package.json:24]: I’ll fix this in a separate PR.

File: tests/commonjs.cjs
Comment (btmills) [inline tests/commonjs.cjs:6]: The `strict` rule knows based on `.eslintrc.cjs` overriding `*.cjs` files' `sourceType` that this file is script source, so it shouldn't complain about this directive.

```suggestion
```

**Review by btmills — APPROVED** [review]
Just one quick inline suggestion. Thanks for all the work converting another package!

**Review by mdjermanovic — APPROVED** [review]
LGTM assuming we'll align the scripts with our release process in a separate PR.

Thanks, @brettz9!

**Comment by nzakas** [comment]
Thanks everyone, I'll merge this and start work on the release scripts. Awesome work @brettz9

**Pull request merged**
